presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizer.java

```diff
@@ -0,0 +1,235 @@
+
+    private final ExecutorService executorService;
+    private final ExecutorService driverService = newFixedThreadPool(1, daemonThreadsNamed("shard-organizer-driver"));
+    private final AtomicBoolean shutdown = new AtomicBoolean();
+
+    private final BlockingQueue<OrganizationSet> queue = new LinkedBlockingQueue<>();
+
+    private final ShardManager shardManager;
+    private final MetadataDao metadataDao;
+    private final ShardCompactor compactor;
+
+    // Tracks shards that are scheduled for compaction so that we do not schedule them more than once
+    private final Set<UUID> shardsInProgress = newConcurrentHashSet();
+
+    private final CounterStat successCount = new CounterStat();
+    private final CounterStat failureCount = new CounterStat();
+
+    @Inject
+    public ShardOrganizer(@ForMetadata IDBI dbi, ShardManager shardManager, ShardCompactor compactor, StorageManagerConfig config)
+    {
+        this(dbi, shardManager, compactor, config.getCompactionThreads());
+    }
+
+    public ShardOrganizer(@ForMetadata IDBI dbi, ShardManager shardManager, ShardCompactor compactor, int threads)
+    {
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.compactor = requireNonNull(compactor, "compactor is null");
+
+        this.executorService = newFixedThreadPool(threads, daemonThreadsNamed("shard-organizer-%s"));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        driverService.submit((Runnable) () -> {
+            while (!Thread.currentThread().isInterrupted() && !shutdown.get()) {
+                try {
+                    OrganizationSet compactionSet = queue.take();
+                    runAsync(new OrganizationJob(compactionSet), executorService)
+                            .whenComplete((none, throwable) -> {
+                                if (throwable == null) {
+                                    successCount.update(1);
+                                }
+                                else {
+                                    log.warn(throwable, "Error in running organization job");
+                                    failureCount.update(1);
+                                }
+                            });
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        });
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        driverService.shutdown();
+        executorService.shutdown();
+    }
+
+    void enqueue(OrganizationSet organizationSet)
+    {
+        shardsInProgress.addAll(organizationSet.getShards());
+        queue.add(organizationSet);
+    }
+
+    boolean inProgress(UUID shardUuid)
+    {
+        return shardsInProgress.contains(shardUuid);
+    }
+
+    public class OrganizationJob
```

Comment:
We don't need to fix this now, but I think this class is mixing the scheduling and the actual jobs. All the fields in `ShardOrganizer` are only needed for the jobs, not the driver. This could be improved by injecting a job factory into `ShardOrganizer`, allowing it to only to know about the factory and the scheduling logic. Then we could test the scheduler by itself, which I'm not sure is tested at all right now.